joblib/test/test_parallel.py

```diff
@@ -753,11 +847,63 @@ def test_memmap_with_big_offset(tmpdir):
 
 def test_warning_about_timeout_not_supported_by_backend():
     with warns(None) as warninfo:
-        Parallel(timeout=1)(delayed(square)(i) for i in range(10))
+        Parallel(timeout=1)(delayed(square)(i) for i in range(50))
     assert len(warninfo) == 1
     w = warninfo[0]
     assert isinstance(w.message, UserWarning)
     assert str(w.message) == (
         "The backend class 'SequentialBackend' does not support timeout. "
         "You have set 'timeout=1' in Parallel but the 'timeout' parameter "
         "will not be used.")
+
+
+@parametrize('backend', ALL_VALID_BACKENDS)
+@parametrize('n_jobs', [1, 2, -2, -1])
+def test_abort_backend(n_jobs, backend):
+    delays = ["a"] + [10] * 100
+
+    if os.environ.get("TRAVIS_OS_NAME") is not None and n_jobs < 0:
+        # Use only up to 8 cpu in travis as cpu_count return 32 whereas we
+        # only access 2 cores.
+        n_jobs += 8
+
+    with raises(TypeError):
+        t_start = time.time()
+        Parallel(n_jobs=n_jobs, backend=backend)(
+            delayed(time.sleep)(i) for i in delays)
+    dt = time.time() - t_start
+    assert dt < 3
+
+
+@with_numpy
+@with_multiprocessing
+@parametrize('backend', PROCESS_BACKENDS)
+def test_memmapping_leaks(backend, tmpdir):
+    # Non-regression test for memmapping backends. Ensure that the data
+    # does not stay too long in memory
+    tmpdir = tmpdir.strpath
+
+    # Use max_nbytes=1 to force the use of memory-mapping even for small
+    # arrays
+    with Parallel(n_jobs=2, max_nbytes=1, backend=backend,
+                  temp_folder=tmpdir) as p:
+        p(delayed(check_memmap)(a) for a in [np.random.random(10)] * 2)
+
+        # The memmap folder should not be clean in the context scope
+        assert len(os.listdir(tmpdir)) > 0
+
+    # Make sure that the shared memory is cleaned at the end when we exit
+    # the context
+    assert not os.listdir(tmpdir)
+
+    # Make sure that the shared memory is cleaned at the end of a call
+    p = Parallel(n_jobs=2, max_nbytes=1, backend=backend)
+    p(delayed(check_memmap)(a) for a in [np.random.random(10)] * 2)
+
+    assert not os.listdir(tmpdir)
+
+
+def test_empty_iterator():
```

Comment:
There is already `test_parallel_with_exhausted_iterator` that does the same thing I think. When I mentioned adding a test for the edge case, I was thinking more of "the first batch was very quick and its callback already dispatched all the remaining jobs".